File: app-java/src/com/nfccopy/app/CardDetailActivity.java
package com.nfccopy.app;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import org.json.JSONObject;
import org.json.JSONException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class CardDetailActivity extends Activity {

    private String cardId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_card_detail);

        cardId = getIntent().getStringExtra("card_id");
        if (cardId == null) { finish(); return; }

        SharedPreferences prefs = getSharedPreferences("nfc_cards", Context.MODE_PRIVATE);
        String json = prefs.getString("card_" + cardId, null);
        if (json == null) { finish(); return; }

        try {
            JSONObject card = new JSONObject(json);
            TextView detailText = (TextView) findViewById(R.id.detailText);
            final Button btnEmulate = (Button) findViewById(R.id.btnEmulate);

            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
            StringBuilder sb = new StringBuilder();
            sb.append("이름: ").append(card.optString("name")).append("\n");
            sb.append("UID: ").append(card.optString("uid")).append("\n");
            sb.append("기술: ").append(card.optString("tech")).append("\n");
            String atqa = card.optString("atqa", "");
            if (!atqa.isEmpty()) sb.append("ATQA: ").append(atqa).append("\n");
            String sak = card.optString("sak", "");
            if (!sak.isEmpty()) sb.append("SAK: ").append(sak).append("\n");
            long time = card.optLong("time", 0);
            if (time > 0) sb.append("읽은 시간: ").append(sdf.format(new Date(time))).append("\n");
            String raw = card.optString("raw", "");
            if (!raw.isEmpty()) {
                sb.append("\n").append(raw);
            }
            detailText.setText(sb.toString());

            String activeId = prefs.getString("active_card", null);
            btnEmulate.setText(cardId.equals(activeId) ? "에뮬레이션 중지" : "에뮬레이션 시작");

            btnEmulate.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    SharedPreferences p = getSharedPreferences("nfc_cards", Context.MODE_PRIVATE);
                    String active = p.getString("active_card", null);
                    if (cardId.equals(active)) {
                        p.edit().remove("active_card").apply();
                        btnEmulate.setText("에뮬레이션 시작");
                        Toast.makeText(CardDetailActivity.this, "에뮬레이션 중지", Toast.LENGTH_SHORT).show();
                    } else {
                        p.edit().putString("active_card", cardId).apply();
                        btnEmulate.setText("에뮬레이션 중지");
                        Toast.makeText(CardDetailActivity.this, "에뮬레이션 시작", Toast.LENGTH_SHORT).show();
                    }
                }
            });
        } catch (JSONException e) {
            finish();
        }
    }
}

File: app/src/main/java/com/nfccopy/app/ui/CardDetailActivity.kt
package com.nfccopy.app.ui

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.nfccopy.app.R
import com.nfccopy.app.data.CardStorage
import com.nfccopy.app.databinding.ActivityCardDetailBinding
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class CardDetailActivity : AppCompatActivity() {

    private lateinit var binding: ActivityCardDetailBinding
    private lateinit var cardStorage: CardStorage

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCardDetailBinding.inflate(layoutInflater)
        setContentView(binding.root)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        cardStorage = CardStorage(this)
        val cardId = intent.getStringExtra("card_id") ?: run {
            finish()
            return
        }

        val card = cardStorage.getCard(cardId) ?: run {
            finish()
            return
        }

        supportActionBar?.title = card.name

        val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())

        val sb = StringBuilder()
        sb.appendLine("${getString(R.string.detail_name)}: ${card.name}")
        sb.appendLine("${getString(R.string.detail_uid)}: ${card.uid}")
        sb.appendLine("${getString(R.string.detail_tech)}: ${card.techList.joinToString(", ")}")
        if (card.atqa.isNotEmpty()) sb.appendLine("ATQA: ${card.atqa}")
        if (card.sak.isNotEmpty()) sb.appendLine("SAK: ${card.sak}")
        sb.appendLine("${getString(R.string.detail_time)}: ${dateFormat.format(Date(card.timestamp))}")

        card.ndef?.let { ndef ->
            sb.appendLine()
            sb.appendLine("=== NDEF ===")
            sb.appendLine("Max Size: ${ndef.maxSize} bytes")
            sb.appendLine("Writable: ${ndef.isWritable}")
            ndef.records.forEachIndexed { i, record ->
                sb.appendLine("Record $i:")
                sb.appendLine("  TNF: ${record.tnf}")
                sb.appendLine("  Type: ${record.type}")
                if (record.payloadText.isNotEmpty()) {
                    sb.appendLine("  Content: ${record.payloadText}")
                }
                sb.appendLine("  Raw: ${record.payload}")
            }
        }

        if (card.rawData.isNotEmpty()) {
            sb.appendLine()
            sb.appendLine("=== ${getString(R.string.detail_raw)} ===")
            sb.append(card.rawData)
        }

        binding.detailText.text = sb.toString()

        binding.btnEmulate.setOnClickListener {
            val activeId = cardStorage.getActiveCardId()
            if (activeId == card.id) {
                cardStorage.clearActiveCard()
                binding.btnEmulate.text = getString(R.string.start_emulation)
            } else {
                cardStorage.setActiveCard(card.id)
                binding.btnEmulate.text = getString(R.string.stop_emulation)
            }
        }

        val isActive = cardStorage.getActiveCardId() == card.id
        binding.btnEmulate.text = if (isActive) getString(R.string.stop_emulation) else getString(R.string.start_emulation)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}
